### DISCUSSION TITLE
Passing values between on_dag_run* on_task_instance_* events inside listener

### DISCUSSION THREAD

**Discussion by mishrasangeeta87** [apache/airflow#30861] [General]
Hi Everyone,

I am using DAG and task  event listeners: https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/example_dags/plugins/event_listener.html .

There are few variables  which I want to propagate **from** _on_dag_run_running_ event **to**  _on_dag_run_success_ event inside the listener. For this purpose I was trying to use the Xcom.set ([xcom.py](https://github.com/apache/airflow/blob/main/airflow/models/xcom.py)) method from inside my on_dag_run_running event.
The Xcom.set method also expects the task_id as a mandatory param which we don't have in our dag_run object. Do let me know how I could pass the variables between the on_dag_run* listener events.

I tried to achieve the same thing with on_task_instance_* by using 
    `XCom.set(key="job_id",value=task_instance.task_id,dag_id=task_instance.dag_id,run_id=task_instance.run_id,task_id=task_instance.task_id)`

 but there also I got the error:
https://gist.github.com/mishrasangeeta87/39b93b8fba2ca4b43f7183a331223227

 Do let me know is it feasible to share some variables among listener events or not.

Thanks,
Sangeeta

**Comment by mishrasangeeta87** [comment]
@mobuchowski Would you be able to help here since this belongs to listeners topic? Last time you were able to solve the query related to listeners.

**Reply by mobuchowski** [reply]
@mishrasangeeta87 If you don't have multiple schedulers, just saving something in some dictionary created in your listener class should probably work. If you do, then maybe Airflow Variables? https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html

**Reply by mobuchowski** [reply]
```
I tried to achieve the same thing with on_task_instance_* by using
XCom.set(key="job_id",value=task_instance.task_id,dag_id=task_instance.dag_id,run_id=task_instance.run_id,task_id=task_instance.task_id)

but there also I got the error:
https://gist.github.com/mishrasangeeta87/39b93b8fba2ca4b43f7183a331223227
```

This error is known and is already fixed on 2.6 branch.

Before that, the workaround is cumbersome and full of dark corners - basically you'd have to run this code in separate thread while waiting on it to complete in the first one. And with that

**Reply by mishrasangeeta87** [reply]
Thank you for the update @mobuchowski .

**Reply by mishrasangeeta87** [reply]
@mobuchowski  I tried to use the variables inside the listener plugin. Initially I got error using the Variable.set method. Searched for the discussions and found https://github.com/apache/airflow/discussions/19744 helpful while setting the values. Updated code based on the inputs: https://gist.github.com/mishrasangeeta87/ecbb5b115207adb0bd50bf3048373bd8

However while reading the values faced the below error:

80+00:00. externally triggered: True>
[2023-04-27 14:56:28,044] {variable.py:272} ERROR - Unable to retrieve variable from secrets backend (MetastoreBackend). Checking subsequent secrets backend.
Traceback (most recent call last):
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/airflow/models/variable.py", line 268, in get_variable_from_secrets
    var_val = secrets_backend.get_variable(key=key)
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/contextlib.py", line 124, in __exit__
    next(self.gen)
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/airflow/utils/session.py", line 36, in create_session
    session.commit()
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1451, in commit
    self._transaction.commit(_to_root=self.future)
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 797, in _prepare_impl
    self.session.dispatch.before_commit(self.session)
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/sqlalchemy/event/attr.py", line 334, in __call__
    fn(*args, **kw)
  File "/Users/sangeetamishra/venv/lib/python3.9/site-packages/airflow/utils/sqlalchemy.py", line 445, in _validate_commit
    raise RuntimeError("UNEXPECTED COMMIT - THIS WILL BREAK HA LOCKS!")
RuntimeError: UNEXPECTED COMMIT - THIS WILL BREAK HA LOCKS!


Do let me know is there any miss doing the Variable.get(), ?

**Reply by mishrasangeeta87** [reply]
I am performing these tasks on local. I have also tried with dictionary in my listener. Apparently keys added across dagrun* events are visible to all the dagrun* events but not to taskinstance events. Similarly keys set inside taskevents* are visible to only taskevents* but not inside the dagrun* events.

**Reply by mobuchowski** [reply]
@mishrasangeeta87 

>Do let me know is there any miss doing the Variable.get(), ?

Can you do the same on getting what you done on setting? Creating new session and passing it?

>I have also tried with dictionary in my listener. Apparently keys added across dagrun* events are visible to all the dagrun* events but not to taskinstance events.

Yeah - they run on totally different Airflow processes, very probably on different machines. What you see is by design.

If you want to generically share information and Variables do not cut it, you might need something external - for example deploy https://redis.io/ or use some object storage like S3 if your data is larger than what Redis can provide.

**Reply by mishrasangeeta87** [reply]
> @mishrasangeeta87
> 
> > Do let me know is there any miss doing the Variable.get(), ?
> 
> Can you do the same on getting what you done on setting?Creating new session and passing it?
> @mobuchowski  Signature of Variable.get doesn't accept the session object


    @classmethod
    def get(
        cls,
        key: str,
        default_var: Any = __NO_DEFAULT_SENTINEL,
        deserialize_json: bool = False,
    ) -> Any:

hence won't be able to pass it.
> > I have also tried with dictionary in my listener. Apparently keys added across dagrun* events are visible to all the dagrun* events but not to taskinstance events.
> 
> Yeah - they run on totally different Airflow processes, very probably on different machines. What you see is by design.
> 
> If you want to generically share information and Variables do not cut it, you might need something external - for example deploy https://redis.io/ or use some object storage like S3 if your data is larger than what Redis can provide.
